src/services/pcn-bridge/src/Bridge_dp_no_stp.c

```diff
@@ -0,0 +1,310 @@
+
+BPF_TABLE("hash", struct port_vlan_key, struct port_vlan_value, port_vlan,
+          1024);
+/*
+ * port_vlan usage example:
+ * <port> <vlan> -> vlanid
+ *
+ * port 1 access vlan=10
+ * 1 VLAN_WILDCARD -> 10
+ *
+ * port 3 trunk vlan=30,40
+ * 3 30 -> 30
+ * 3 40 -> 40
+ *
+ * port 4 trunk vlan=all
+ * 4 VLAN_WILDCARD -> VLAN_WILDCARD
+ */
+
+/*
+ * Fowarding Table: VLAN + MAC Address to port association.  This table is
+ * filled
+ * in the learning phase and then is used in the forwarding phase to decide
+ * where to send a packet
+ */
+struct fwd_key {
+  u16 vlan;
+  __be64 mac;
+} __attribute__((packed));
+
+struct fwd_entry {
+  u64 timestamp;
+  u32 port;
+  enum entry_type type;
+} __attribute__((packed));
+
+BPF_TABLE("hash", struct fwd_key, struct fwd_entry, fwdtable, 1024);
+
+/*
+ * Port Table: Saves the state of the ports
+ * TODO: what is the best option in this case, hash vs array?
+ */
+struct port {
+  u16 mode;
+  u16 native_vlan;
+  bool native_vlan_enabled;
+};
+
+BPF_TABLE("hash", u32, struct port, ports, 256);
+
+struct eth_hdr {
+  __be64 dst : 48;
+  __be64 src : 48;
+  __be16 proto;
+} __attribute__((packed));
+
+static int handle_rx(struct CTXTYPE *ctx, struct pkt_metadata *md) {
+  void *data = (void *)(long)ctx->data;
+  void *data_end = (void *)(long)ctx->data_end;
+  struct eth_hdr *eth = data;
+
+  if (data + sizeof(*eth) > data_end)
+    goto DROP;
+
+  u32 in_ifc = md->in_port;
+
+  pcn_log(ctx, LOG_DEBUG, "in_ifc=%d proto=0x%x", in_ifc,
+          bpf_ntohs(eth->proto));
+  pcn_log(ctx, LOG_DEBUG, "mac src:%M dst:%M", eth->src, eth->dst);
+
+  u16 vlanid = 0;
+#ifdef POLYCUBE_XDP
+  bool tagged = pcn_is_vlan_present(ctx);
+
+  if (tagged) {
+    u16 eth_proto;
+    if (pcn_get_vlan_id(ctx, &vlanid, &eth_proto) < 0)
+      goto DROP;
+  }
+#else
+  bool tagged = ctx->vlan_present;
+
+  if (tagged)
+    vlanid = ctx->vlan_tci & 0x0fff;
+#endif
+
+  struct port *in_port = ports.lookup(&in_ifc);
+  if (!in_port) {
+    pcn_log(ctx, LOG_ERR, "Port %u not found", in_ifc);
+    goto DROP;
+  }
+
+  struct port_vlan_key vlan_key;
+  struct port_vlan_value *vlan_entry;
+
+  vlan_key.port = in_ifc;
+
+  if (in_port->mode == PORT_MODE_ACCESS) {
+    if (tagged)
+      goto DROP;
+
+    vlan_key.vlan = VLAN_WILDCARD;
+    vlan_entry = port_vlan.lookup(&vlan_key);
+    if (!vlan_entry)
+      goto DROP;  // TODO: is this code correct?
+
+    vlanid = vlan_entry->vlan;
+
+    pcn_log(ctx, LOG_TRACE, "port mode ACCESS vlan=%d, state=%d", vlanid,
+            vlan_entry->stp_state);
+  } else {
+    // TRUNK
+    if (!tagged) {
+      vlanid = in_port->native_vlan;
+    }
+    vlan_key.vlan = vlanid;
+    vlan_entry = port_vlan.lookup(&vlan_key);
+    if (!vlan_entry) {
+      // TODO manage case of all vlans allowed
+      goto DROP;
+    }
+  }
+
+  // LEARNING PHASE
+  struct fwd_key src_key;
+  src_key.vlan = vlanid;
+  src_key.mac = eth->src;
+  struct fwd_entry e;  // used below
+  e.timestamp = bpf_ktime_get_ns();
```

Comment:
Can we  use the other primitive that returns the time with the precision of one second? This call is definitely costly.

Reply:
It is not available in the framework yet: https://github.com/polycube-network/polycube/issues/58.
I think the only service using the "one second precision" is iptables.

Reply:
@frisso Actually in simplebridge service i noticed that there was the thread you mentioned, but i did not used it in the bridge because honestly i thought it was useless (now i understand why it was used!).
I could implement the same strategy here, if you wish.

Reply:
It would be very nice. Keeping the same strategies across all services makes our framework better and more organic. Thanks!